File: leetCode1584.py
class UnionFind:
    def __init__(self, numOfNodes):
        self.parent = self.makeSet(numOfNodes)
        self.size = [1 for _ in range(numOfNodes)]
    
    def makeSet(self, numOfNodes):
        return [x for x in range(numOfNodes)]
    
    def find(self, node):
        while node != self.parent[node]:
            self.parent[node] = self.parent[self.parent[node]]
            node = self.parent[node]
        return node
    
    def union(self, node1, node2):
        root1 = self.find(node1)
        root2 = self.find(node2)
        if root1 == root2:
            return False
     
        if self.size[root1] > self.size[root2]:
            self.parent[root2] = root1
            self.size[root1] += 1
        else:
            self.parent[root1] = root2
            self.size[root2] += 1
                
        return True
        
        
class Solution(object):
    def minCostConnectPoints(self, points):
        """
        :type points: List[List[int]]
        :rtype: int
        """
        numOfNodes = len(points) 
        edges = []
        for i in range(numOfNodes):
            for j in range(i+1, numOfNodes):
                distance = abs(points[i][0]-points[j][0]) + abs(points[i][1]-points[j][1])
                edges.append((distance, i, j))
        
        # sort based on distance
        edges.sort()
        
        # Kruskal's algorithm
        cost = 0
        edgeCount = 0
        
        uf = UnionFind(numOfNodes)
        for distance, node1, node2 in edges:
            if uf.union(node1, node2):
                cost += distance
                edgeCount += 1
                if edgeCount == numOfNodes - 1:
                    return cost
        
        return cost
            
if __name__ == "__main__":
    points = [[0,0],[2,2],[3,10],[5,2],[7,0]]
    solution = Solution()
    print(solution.minCostConnectPoints(points))

# output: 20

File: leetCode1584.go
package main

import (
	"fmt"
	"sort"
)

type UnionFind struct {
	parent []int
	size   []int
}

func newUnionFind(numOfNodes int) *UnionFind {
	// makeSet
	parent := make([]int, numOfNodes)
	size := make([]int, numOfNodes)
	for i := 0; i < numOfNodes; i++ {
		parent[i] = i
		size[i] = 1
	}
	return &UnionFind{
		parent: parent,
		size:   size,
	}
}

func (uf *UnionFind) find(node int) int {
	for node != uf.parent[node] {
		uf.parent[node] = uf.parent[uf.parent[node]]
		node = uf.parent[node]
	}
	return node
}

func (uf *UnionFind) union(node1, node2 int) bool {
	root1 := uf.find(node1)
	root2 := uf.find(node2)

	if root1 == root2 {
		return false
	}

	if uf.size[root1] > uf.size[root2] {
		uf.parent[root2] = root1
		uf.size[root1] += 1
	} else {
		uf.parent[root1] = root2
		uf.size[root2] += 1
	}
	return true
}

type edgeInfo struct {
	distance int
	node1    int
	node2    int
}

func minCostConnectPoints(points [][]int) int {
	numOfNodes := len(points)

	edges := []*edgeInfo{}
	for i := 0; i < numOfNodes; i++ {
		for j := i + 1; j < numOfNodes; j++ {
			distance := abs(points[i][0]-points[j][0]) + abs(points[i][1]-points[j][1])
			edges = append(edges, &edgeInfo{distance: distance, node1: i, node2: j})
		}
	}

	sort.Slice(edges, func(i, j int) bool {
		return edges[i].distance < edges[j].distance
	})

	uf := newUnionFind(numOfNodes)

	cost := 0
	edgeCount := 0
	for _, edge := range edges {
		if uf.union(edge.node1, edge.node2) {
			cost += edge.distance
			edgeCount += 1
			if edgeCount == numOfNodes-1 {
				return cost
			}
		}
	}
	return cost
}

func abs(num int) int {
	if num < 0 {
		return -num
	}
	return num
}

func main() {
	points := [][]int{
		{0, 0},
		{2, 2},
		{3, 10},
		{5, 2},
		{7, 0},
	}

	fmt.Println(minCostConnectPoints(points))
}

// output: 20
